### ISSUE TITLE
[HOLD for payment 2024-07-10] [$250] [Wave Collect] Disable all but the default distance rate when feature is disabled same for categories and tags

### ISSUE THREAD

**Issue by mountiny** [Expensify/App#39463]
### Problem

When the Distance Rates feature is disabled in NewDot, the distance rates remain enabled. Similarly to other features in the App, we should disable the rates when the feature is disabled.

However, this is also not yet implemented optimistically for tags and categories. This is an edge case that helps in cases of rare offline usage, and as such, it was not noted as regression. 

When we disable the categories and tags feature in the backend, we also disable any enabled categories and tags. We need to do the same optimistically in the client (failure data should revert the state to the one before toggling the feature off). 

### Solution

When the distance rates feature is turned off in NewDot, go through all the distance rates on the policy and disable them except the `Default Rate` if available. If it's not available, keep the first rate that you go over as enabled and disable all the rest. We need to keep one enabled.

For tags and categories, make sure to turn off the `requiresCategory` and `requiresTag` optimistically and then all the tags, even multi-level.

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010ba691024d0e69b3</li>
        <li>Upwork Job ID: 1775299514448031744</li>
        <li>Last Price Increase: 2024-06-07</li>
<li>Automatic offers: </li>
<ul>
<li>getusha | Reviewer | 102703385</li>
<li>dominictb | Contributor | 102703388</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @kadiealexander</details>

**Comment by mountiny** [comment]
Will try to look into this tomorrow

**Comment by trjExpensify** [comment]
>Similarly to other features in the App, we should disable the rates when the feature is disabled (we disable all categories, tags and taxes when the feature is disabled)

Apart from the default or whatever that is, right?

**Comment by mountiny** [comment]
working on this one today

**Comment by mountiny** [comment]
adding tests

**Comment by mountiny** [comment]
PR is ready for a review

**Comment by mountiny** [comment]
Also: lets make sure the App handles this optimistically

**Comment by mountiny** [comment]
Will have to handle this later this week, focusing on more urgent tasks

**Comment by mountiny** [comment]
Will look into this soon

**Comment by mountiny** [comment]
What is left, when disabling the distance feature, optimistically disable the distance rates on the policy

**Comment by mountiny** [comment]
Back on this week, focusing on performance and quality tasks

**Comment by mountiny** [comment]
Focusing on 1:1:1 tasks

**Comment by mountiny** [comment]
Still on my list for this week

**Comment by mountiny** [comment]
Will look tomorrow

**Comment by mountiny** [comment]
No I see we are not doing this for tags or categories optimistically either. 

I will update this task to be external and expand the scope to categories and tags too

**Comment by mountiny** [comment]
Discussing internally the next steps

**Comment by mountiny** [comment]
Looking for proposals

**Comment by b4s36t4** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Disable all but the default distance rate when feature is disabled same for categories and tags
### What is the root cause of that problem?

NA - Features

### What changes do you think we should make in order to solve the problem?

[https://github.com/Expensify/App/blob/a51a245ac4a3f4cae0635e1564d530b6c186e3c4/src/libs/actions/Policy/Policy.ts#L4389 here in this function we disable or enable the feature, when the feature](https://github.com/Expensify/App/blob/e9dc2ef95eb8007aa2b41f676fe952da0359ef47/src/pages/workspace/WorkspaceMoreFeaturesPage.tsx#L66)

Here in this page we call the enable/disable feature flag, we can add the logic here for the disabling the customUnits.

Logic for disabling the custom unit rates, we should update the logic here https://github.com/Expensify/App/blob/e9dc2ef95eb8007aa2b41f676fe952da0359ef47/src/pages/workspace/WorkspaceMoreFeaturesPage.tsx#L64

```jsx
if (!isEnabled) {
    Object.entries(policyCustomUnit?.rates ?? {}).forEach(([key, value], index) => {
        if (value.name === 'Default Rate' || index === 0) {
            return;
        }

        return (value.enabled = false);
    });
    // Policy.setPolicyDistanceRatesEnabled(policy?.id ?? '', policyCustomUnit, Object.values(policyCustomUnit.rates));
}
Policy.enablePolicyDistanceRates(policy?.id ?? '', isEnabled, policyCustomUnit?.customUnitID, policyCustomUnit);
```

This way we can send the updated custom unit into the function we're calling i.e `enablePolicyDistanceRates`.

We need to extend the parameters and `VALUE` object for `optimisticData` variable inside the function to enable passing the data to policy.

Need to update the logic in this file https://github.com/Expensify/App/blob/a51a245ac4a3f4cae0635e1564d530b6c186e3c4/src/libs/actions/Policy/Policy.ts#L3671

```js
function enablePolicyDistanceRates(policyID: string, enabled: boolean, customUnitID?: string, customUnit?: CustomUnit | null) {
```

Update the `optimisticData` key's value with the following change.

```js
...(customUnit && customUnitID ? {customUnits: {[customUnitID]: customUnit}} : {}),
```



### What alternative solutions did you explore? (Optional)
NA

**Comment by mountiny** [comment]
@getusha ready for you to review the proposals.

**Comment by getusha** [comment]
Will update today

**Comment by getusha** [comment]
>When the Distance Rates feature is disabled in NewDot, the distance rates remain enabled. Similarly to other features in the App, we should disable the rates when the feature is disabled.

@mountiny i think this is already implemented, same behavior on offline.

https://github.com/Expensify/App/assets/75031127/c6f7550d-0353-42a0-96f9-10cec0f31889

**Comment by mountiny** [comment]
I dont think it can work when you are offline, can you check again please? there is no code that would do this optimistically, is there?

**Comment by mountiny** [comment]
@getusha bump for the above please

**Comment by getusha** [comment]
@b4s36t4's detailed [proposal](https://github.com/Expensify/App/issues/39463#issuecomment-2142806097) looks good to me.
There are some issues i noticed after applying the solution, but i think we can address them in the PR.
🎀 👀 🎀 C+ Reviewed.

**Comment by b4s36t4** [comment]
I disagree with you, it's stated that there shouldn't be code diffs, I added a code block here explaining what parts of applications need to be updated. Everyone has different ways of understanding and explaining you explained in plain english, I explained it using Code, that doesn't mean or give you advantage that your proposal should be accepted. 

> Do similar steps for categories and tags, make sure it works for multi-level categories/tags too, and also turn off the requiresCategory and requiresTag optimistically. It's possible that we don't need to leave any tag/category enabled (like distance rate), can just disable all of them.

Similar steps, You just published your proposal with limited information, I provided my proposal with more information so that reviewer get's what I'm saying.

Also I believe you only wanted your proposal to be first, that's you posted your proposal without any info at first edit and then you made edits going on, doesn't that discouraged? 


What I'm trying to say is that everyone has their own way of doing things, you can defend your proposal when I copy your solution or when my proposal can cause issues or not a best practice. At this instance you don't have a say to defend your proposal.

Again sorry if my words are a bit harsh, thanks!

**Comment by mountiny** [comment]
I agree we should make sure we explain the proposal in plain English based on the contributing guidelines so assigning @dominictb 

I understand its easier for the C+ to test fully but also ideally we do not have to do that before the proposal is accepted, the explanation in plain English should prove deep understanding of the domain and that even if there would be some hurdles along the way in implementation phase, we can overcome them

**Comment by b4s36t4** [comment]
@mountiny plain English does mean the proposal should be understandable not to put things in plain English right? Whatever it is there is not at all covering what he said. 

He just explained similar steps we have to do with categories and tags which is not at all understandable. 

How would that get's accepted?

**Comment by mountiny** [comment]
The idea is that, ad absurdum, you could already just post your branch with a solution and the full code diff and obviously, that would be the easiest as we can just go straight up to test it. But we want to make sure we can explain in words what is happening and why it's happening and how we want to fix it. 

Again, this is hard, and I am sorry; this time, it went on the other side, and next time, it will go for you. When the proposals are so similar its hard to pick one with clear reasons

**Comment by mountiny** [comment]
this should be ready for payment when the time comes 

$250 to @dominictb and to @getusha

**Comment by kadiealexander** [comment]
@getusha @mountiny do we need a regession test here?

**Comment by kadiealexander** [comment]
The contracts expired so I've resent them here. Payouts due:

- [ ] Contributor: $250 @dominictb ([offer](https://www.upwork.com/nx/wm/offer/103117716))
- [ ] Contributor+: $250 @getusha ([offer](https://www.upwork.com/nx/wm/offer/103117703))

Upwork job is [here](https://www.upwork.com/jobs/~01973a18bf317bf8a4).

**Comment by kadiealexander** [comment]
Thanks @dominictb, I've updated the offers.

**Comment by kadiealexander** [comment]
@getusha please share the regression test steps.

**Comment by getusha** [comment]
>[@getusha] Please propose [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the new feature will work correctly on production in further releases.

**Regression Test Proposal**

1. Go offline
2. Go to workspaces > Select a workspace
3. Go to categories
4. Make sure to have one category enabled
5. Go to more features
6. Disable & Enable `Categories` feature
7. Go to categories
8. Verify that every category is disabled
9. Go to distance rates
10. Make sure to have at least two rates enabled
11. Go to more features page
12. Disable & Enable `distance rates` feature
13. Go to `Distance rates` page
14. Verify that only one distance rate remains enabled
15. Go to `Tags` page
16. Make sure to have at least one tag enabled
17. Go to more features page
18. Disable & Enable `Tags` feature
19. Go to `Tags` page
20. Verify that every options are disabled

Do we agree 👍 or 👎
cc @mountiny @dominictb

**Comment by getusha** [comment]
Accepted the offer @kadiealexander
